daemon/cmd/ipcache.go

```diff
@@ -87,3 +104,126 @@ func containsSubnet(outer, inner net.IPNet) bool {
+// that ensures the prefix exists and it gets the same numeric identity before. It
+// reserves those local numeric IDs from allocation to ensure they are available.
+//
+// For ingress IPs, it will add those to the ipcache and configure the local node
+// accordingly.
+//
+// This *must* be called before initMaps(), which will clobber the "old" ipcache.
+func (d *Daemon) restoreIPCache() error {
+	ingressIPs := make([]netip.Prefix, 0, 2)
+	// need to preserve this so we can remove it later.
+	d.restoredCIDRs = map[netip.Prefix]identity.NumericIdentity{}
+
+	// Dump the bpf ipcache, recording any prefixes with local or ingress
+	// numeric identities.
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+
+		if isLocalIdentity(nid) {
+			d.restoredCIDRs[k.Prefix()] = nid
+		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
+			ingressIPs = append(ingressIPs, k.Prefix())
+		}
+	})
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error dumping ipcache: %w", err)
+	}
+	// DumpWithCallback() leaves the ipcache map open, must close before opened for
+	// parallel mode in Daemon.initMaps()
+	ipcachemap.IPCacheMap().Close()
+
+	// Now that the map is dumped,
+	// - upsert relevant metadata in to the ipcache
+	// - reserve all existing CIDR identities
+	// - add Ingress IPs to local Node.
+	metaUpdates := make([]ipcache.MU, 0, len(ingressIPs)+len(d.restoredCIDRs))
+	nidsToReserve := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+
+	for prefix, nid := range d.restoredCIDRs {
+		nidsToReserve = append(nidsToReserve, nid)
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
+		})
+	}
+	for _, prefix := range ingressIPs {
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: ipcachetypes.NewResourceID(ipcachetypes.ResourceKindDaemon, "", "ingress"),
+			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+		})
+
+		// Set any restored ingress IPs back on the LocalNode object
+		d.nodeLocalStore.Update(func(n *node.LocalNode) {
+			addr := prefix.Addr()
+			if addr.Is4() {
+				n.IPv4IngressIP = addr.AsSlice()
+			} else {
+				n.IPv6IngressIP = addr.AsSlice()
+			}
+		})
+	}
+
+	// Insert the batched changes in to the ipcache.
+	// Even though the ipcache map hasn't been initialized yet, this is
+	// safe to do so, because the ipcache's apply controller is currently
+	// paused.
+	d.ipcache.IdentityAllocator.ReserveIdentities(nidsToReserve)
+	d.ipcache.UpsertMetadataBatch(metaUpdates...)
+
+	return nil
+}
+
+// releaseRestoredCIDRS removes the "stub" metadata that was inserted
```

Comment:
Oh yeah, good point. "stub" isn't even that useful a term, since it's valid metadata like anything else. Perhaps "placeholder" is a better term. I'll write some more words in the docblock

Reply:
Renamed this from "stub" to "placeholder" and tried to give some more context in the documentation.